File: src/python/popularSuperHero.py
from pyspark import SparkConf, SparkContext


def countCoOccurences(line):
    fields = line.split()
    return fields[0], (len(fields)-1)


def loadHeros(line):
    fields = line.split('\"')
    return int(fields[0]), fields[1]


conf = SparkConf().setMaster("local").setAppName("PopularSuperHero")
sc = SparkContext(conf=conf)

heros = sc.textFile('./Marvel-Names.txt')
heroNames = heros.map(loadHeros)

lines = sc.textFile('./Marvel-Graph.txt')
coOccurences = lines.map(countCoOccurences)
total = coOccurences.reduceByKey(lambda x, y: x+y)
flipped = total.map(lambda x: (x[1], x[0]))
mostPopular = flipped.max()
mostPopularName = heroNames.lookup(int(mostPopular[1]))[0]
print(mostPopularName + ' with ' + str(mostPopular[0]) + ' co-appearances.')

File: src/scala/PopularSuperHero.scala
import org.apache.spark.SparkContext
import org.apache.spark.SparkConf

object PopularSuperHero{
  def main(args: Array[String]): Unit = {
    var conf = new SparkConf().setMaster("local").setAppName("Popular SeperHero")
    var sc = new SparkContext(conf)

    var heroFile = sc.textFile("./Marvel-Names.txt")
    var heroNames = heroFile.map(loadHeroes)

    var textFile = sc.textFile("./Marvel-Graph.txt")
    var coOccurences = textFile.map(countCoOccurences)
    var total = coOccurences.reduceByKey((x, y) => x+y)
    var flipped = total.map(x => (x._2, x._1)).sortByKey()
    var mostPopular = flipped.max()
    var mostPopularName = heroNames.lookup(mostPopular._2.toInt)(0)
    println(mostPopularName)
  }

  def countCoOccurences(line: String): (String, Int) = {
    var fields = line.split(" ")
    return (fields(0), fields.length-1)
  }

  def loadHeroes(line: String): (Int, String) = {
    var fields = line.split("\"")
    var heroId = fields(0).split(" ")
    if(fields.length>1)
      return (heroId(0).toInt, fields(1))
    else
      return (heroId(0).toInt, "")
  }
}
